src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,694 @@
+                    {
+                        min_stale_id = std::cmp::min(min_stale_id, seq);
+                        max_stale_id = std::cmp::max(max_stale_id, seq);
+                    }
+                }
+            }
+        });
+        let mut files: VecDeque<FileWithFormat<F>> = VecDeque::default();
+        if max_stale_id > 0 {
+            files.reserve((max_stale_id - min_stale_id) as usize + 1);
+            for seq in min_stale_id..=max_stale_id {
+                let file_id = FileId {
+                    queue: LogQueue::Append,
+                    seq,
+                };
+                let path = file_id.build_stale_file_path(path);
+                if !path.exists() {
+                    files.clear();
+                } else {
+                    let handle = Arc::new(file_system.open(&path)?);
+                    // It's not necessary to record the metadata of stale files.
+                    file_system.delete_metadata(&path)?;
+                    files.push_back(FileWithFormat {
+                        handle,
+                        format: LogFileFormat::new(Version::V2, 0 /* alignment */),
+                        path_id: 0,
+                    });
+                }
+            }
+            first_seq = max_stale_id - files.len() as FileSeq + 1;
+        }
+        Ok(FileList {
+            first_seq,
+            fds: files,
+        })
+    }
+
+    /// Generates a Fake log used for recycling.
+    ///
+    /// Attention, this function is only called when `Config.enable-log-recycle`
+    /// is true.
+    fn build_stale_file(
+        file_system: &F,
+        paths: &Paths,
+        file_id: FileId,
+        format: LogFileFormat,
+        target_file_size: usize,
+    ) -> Result<FileWithFormat<F>> {
+        let path_id = FileCollection::<F>::get_valid_path(paths, target_file_size);
+        let file_path = file_id.build_stale_file_path(&paths[path_id]);
+        let fd = Arc::new(file_system.create(&file_path)?);
+        let mut writer = file_system.new_writer(fd.clone())?;
+        let mut written = 0_usize;
+        let buf = vec![0; std::cmp::min(LOG_STALE_FILE_BUF_SIZE, target_file_size)];
+        while written <= target_file_size {
+            writer.write_all(&buf).map_err(|e| {
+                writer
+                    .seek(SeekFrom::Start(written as u64))
+                    .unwrap_or_else(|e| {
+                        panic!("failed to reseek after write failure: {}", e);
+                    });
+                e
+            })?;
+            written += buf.len();
+        }
+        // Metadata of stale files are not what we're truely concerned. So,
+        // they can be ignored by clear them here.
+        Ok(FileWithFormat {
+            handle: fd,
+            format,
+            path_id,
+        })
+    }
+
+    /// Clear a log file by the given metadata.
+    fn clear_file(
```

Comment:
Doesn't make sense to wrap such a small function inside another function, especially when it's behavior is controlled by a boolean, which is a bad practice.

If you want to fail point the delete operation, just do it inside `FileSystem`.